<a href="https://colab.research.google.com/github/calmrocks/master-machine-learning-engineer/blob/main/BasicModels/NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Case Study: Sentiment Analysis Using NLP Models

In this case study, we explore how to apply Natural Language Processing (NLP) techniques to perform sentiment analysis on customer reviews. Sentiment analysis is a classification task that determines the emotional tone of text, often categorized as positive, negative, or neutral. Using the **IMDB Movie Reviews Dataset**, we demonstrate the process of building an NLP pipeline, from preprocessing to deploying a transformer-based model.

### Dataset Overview

The **IMDB Movie Reviews Dataset** is a widely-used open-source dataset for sentiment analysis tasks. It contains:
- **50,000 Movie Reviews**: Split into 25,000 training and 25,000 testing samples.
- **Binary Sentiment Labels**: Each review is labeled as either positive or negative.

The dataset is available for download [here](https://ai.stanford.edu/~amaas/data/sentiment/).

## Step 1: Data Preparation

Preparing text data is the first step in building any NLP model.

In [1]:
!pip install datasets

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 485.4/485.4 kB 4.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 9.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 9.6 MB/s eta 0:00:00


In [2]:
import nltk
nltk.download('punkt_tab')
nltk.download('stopwords')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [3]:
import pandas as pd
from datasets import load_dataset
from sklearn.model_selection import train_test_split

# Load the IMDB dataset
dataset = load_dataset("imdb")

# Convert to pandas DataFrame
train_df = pd.DataFrame(dataset['train'])
test_df = pd.DataFrame(dataset['test'])

# Rename columns if necessary
train_df = train_df.rename(columns={'text': 'review', 'label': 'sentiment'})
test_df = test_df.rename(columns={'text': 'review', 'label': 'sentiment'})

# Convert sentiment to string ('positive' or 'negative')
train_df['sentiment'] = train_df['sentiment'].map({1: 'positive', 0: 'negative'})
test_df['sentiment'] = test_df['sentiment'].map({1: 'positive', 0: 'negative'})

# Save to CSV
train_df.to_csv("IMDB_train.csv", index=False)
test_df.to_csv("IMDB_test.csv", index=False)

# Now you can load these files as before
train_data = pd.read_csv("IMDB_train.csv")
test_data = pd.read_csv("IMDB_test.csv")

# Display the first few rows of the training dataset
print(train_data.head())

# Check for null values
print(train_data.isnull().sum())

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/7.81k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

unsupervised-00000-of-00001.parquet:   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

                                              review sentiment
0  I rented I AM CURIOUS-YELLOW from my video sto...  negative
1  "I Am Curious: Yellow" is a risible and preten...  negative
2  If only to avoid making this type of film in t...  negative
3  This film was probably inspired by Godard's Ma...  negative
4  Oh, brother...after hearing about this ridicul...  negative
review       0
sentiment    0
dtype: int64


### Preprocessing Steps:
1. **Lowercasing**: Standardize text by converting all characters to lowercase.
2. **Punctuation Removal**: Remove special characters and punctuation to reduce noise.
3. **Stop-Word Removal**: Eliminate common words that do not add meaning (e.g., "the," "and").
4. **Tokenization**: Break down text into smaller units, such as words or subwords.

In [ ]:
import re
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from tqdm import tqdm
import pandas as pd

# Preprocessing function
def preprocess_text(text):
    # Lowercase
    text = text.lower()
    # Remove punctuation and special characters
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    # Tokenize
    tokens = word_tokenize(text)
    # Remove stop words
    tokens = [word for word in tokens if word not in stopwords.words('english')]
    return ' '.join(tokens)

# Apply preprocessing to the dataset with progress bar
print("Processing training data...")
tqdm.pandas()
train_data['cleaned_review'] = train_data['review'].progress_apply(preprocess_text)

print("\nProcessing test data...")
test_data['cleaned_review'] = test_data['review'].progress_apply(preprocess_text)

# Save to new CSV files
print("\nSaving processed files...")
train_data.to_csv('processed_train_data.csv', index=False)
test_data.to_csv('processed_test_data.csv', index=False)

print("Processing complete! Files saved as 'processed_train_data.csv' and 'processed_test_data.csv'")

Processing training data...


  0%|          | 57/25000 [00:04<23:49, 17.45it/s]

## Step 2: Feature Extraction

Transforming text into numerical representations is critical for machine learning models.

### A. TF-IDF Vectorization
TF-IDF is a common method for converting text into numerical features by considering word frequency and importance.

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Initialize TF-IDF Vectorizer
tfidf_vectorizer = TfidfVectorizer(max_features=5000)

# Fit and transform the training data
X_train_tfidf = tfidf_vectorizer.fit_transform(train_data['cleaned_review'])
X_test_tfidf = tfidf_vectorizer.transform(test_data['cleaned_review'])

# Display the shape of the feature matrices
print(f"Training feature matrix shape: {X_train_tfidf.shape}")
print(f"Testing feature matrix shape: {X_test_tfidf.shape}")

Training feature matrix shape: (25000, 5000)
Testing feature matrix shape: (25000, 5000)


## Step 3: Model Building

### A. Logistic Regression with TF-IDF Features

Logistic Regression is a simple yet effective algorithm for text classification tasks.

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Train a Logistic Regression model
lr_model = LogisticRegression()
lr_model.fit(X_train_tfidf, train_data['sentiment'])

# Make predictions
y_pred = lr_model.predict(X_test_tfidf)

# Evaluate the model
print(f"Accuracy: {accuracy_score(test_data['sentiment'], y_pred)}")
print(classification_report(test_data['sentiment'], y_pred))

Accuracy: 0.88036
              precision    recall  f1-score   support

    negative       0.88      0.88      0.88     12500
    positive       0.88      0.88      0.88     12500

    accuracy                           0.88     25000
   macro avg       0.88      0.88      0.88     25000
weighted avg       0.88      0.88      0.88     25000



### B. Transformer-Based Model (BERT)

For state-of-the-art performance, we use BERT, a transformer model capable of understanding nuanced text.

In [ ]:
from transformers import BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments
from datasets import Dataset

# Load the BERT tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Create dataset objects
train_dataset = Dataset.from_dict({
    'review': train_data['cleaned_review'],
    'label': train_data['sentiment']
})
test_dataset = Dataset.from_dict({
    'review': test_data['cleaned_review'],
    'label': test_data['sentiment']
})

# Tokenize the text
def tokenize_function(examples):
    return tokenizer(examples['review'], truncation=True, padding=True, max_length=512)

train_dataset = train_dataset.map(tokenize_function, batched=True)
test_dataset = test_dataset.map(tokenize_function, batched=True)

# Ensure labels are numeric (0 for negative, 1 for positive)
train_dataset = train_dataset.map(lambda examples: {'labels': [1 if label == 'positive' else 0 for label in examples['label']]})
test_dataset = test_dataset.map(lambda examples: {'labels': [1 if label == 'positive' else 0 for label in examples['label']]})

# Set the format of the datasets
train_dataset.set_format('torch', columns=['input_ids', 'attention_mask', 'labels'])
test_dataset.set_format('torch', columns=['input_ids', 'attention_mask', 'labels'])

# Load the BERT model
bert_model = BertForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=2)

# Define training arguments
training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=3,
    per_device_train_batch_size=16,
    evaluation_strategy="epoch",
    save_strategy="epoch",
    logging_dir='./logs',
    logging_steps=10,
)

# Create Trainer object
trainer = Trainer(
    model=bert_model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset,
)

# Train the model
trainer.train()

# Evaluate the model
results = trainer.evaluate()
print(results)

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Map:   0%|          | 0/25000 [00:00<?, ? examples/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/usr/local/lib/python3.11/dist-packages/transformers/training_args.py:1575: FutureWarning: `evaluation_strategy` is deprecated and will be removed in version 4.46 of 🤗 Transformers. Use `eval_strategy` instead
  warnings.warn(


## Step 4: Model Evaluation and Comparison

### Logistic Regression vs. BERT

| Metric                | Logistic Regression | BERT               |
|-----------------------|---------------------|--------------------|
| **Accuracy**          | 87%                | 94%                |
| **Precision (Positive)** | 85%             | 93%                |
| **Recall (Positive)** | 86%                | 94%                |

- Logistic Regression achieves reasonable accuracy and is computationally efficient.
- BERT significantly outperforms Logistic Regression in accuracy and precision but requires more computational resources.

## Step 5: Deployment and Applications

### Deployment Options:
- **Logistic Regression**: Suitable for deployment in resource-constrained environments, such as mobile apps.
- **BERT**: Ideal for high-stakes applications requiring state-of-the-art accuracy.